sdk/python/kfp/dsl/for_loop_test.py

```diff
@@ -89,12 +98,44 @@ def test_get_subvar_type(self, dict_type, value_type):
                 '{{channel:task=task1;name=output1-loop-item;type=Dict[str, str];}}',
         },
     )
-    def test_loop_argument_from_pipeline_channel(self, channel,
-                                                 expected_serialization_value):
-        loop_argument = for_loop.LoopArgument.from_pipeline_channel(channel)
+    def test_loop_parameter_argument_from_pipeline_channel(
+            self, channel, expected_serialization_value):
+        loop_argument = for_loop.LoopParameterArgument.from_pipeline_channel(
+            channel)
         self.assertEqual(loop_argument.items_or_pipeline_channel, channel)
         self.assertEqual(str(loop_argument), expected_serialization_value)
 
+    @parameterized.parameters(
+        {
+            'channel':
+                pipeline_channel.PipelineArtifactChannel(
+                    name='param1',
+                    channel_type='system.Artifact@0.0.1',
+                    task_name='task1',
+                    is_artifact_list=False,
+                ),
+            'expected_serialization_value':
+                '{{channel:task=task1;name=param1-loop-item;type=system.Artifact@0.0.1;}}',
+        },
+        {
+            'channel':
+                pipeline_channel.PipelineArtifactChannel(
+                    name='output1',
+                    channel_type='system.Dataset@0.0.1',
+                    task_name='task1',
+                    is_artifact_list=False,
```

Comment:
Should `is_artifact_list` be `True` since we are going to iterate over it? If yes, should we add validation to catch the case where `is_artifact_list=False`?

As a concrete example, this pipeline happens to compile, but it should not:

```python
@dsl.pipeline
def my_pipeline():
    single_artifact = make_single_artifact(content='foo')
    with dsl.ParallelFor(items=single_artifact.output) as item:
        print_artifact(artifact=item)
```

Note that it does run successfully due to an implementation detail of how we implement artifacts, but we should block this on the compiler-side since it doesn't make sense to iterate over a single `Artifact`.

Reply:
Done, also noticed same issue affects parameters. I was able to compile the corresponding parameter pipeline, but it failed without a clear error. Will address this is the next PR.


```
@dsl.component
def str_identity(s: str) -> str:
    return s


@dsl.pipeline
def my_pipeline():
single_param = str_identity(s='string')
    with dsl.ParallelFor(items=single_param.output) as item:
        str_identity(s=item)
```